File: ForgeHub/server/Api/AuthApi.cs
using ForgeHub.Auth;
using ForgeHub.Data;

namespace ForgeHub.Api;

public static class AuthApi
{
    public static void MapAuth(this WebApplication app)
    {
        app.MapGet("/api/health", () => Results.Ok(new
        {
            product = "ForgeHub",
            status = "OK",
            time = Sql.Now(),
        }));

        // --------------------------------------------------------------
        app.MapPost("/api/auth/login", async (HttpContext context) =>
        {
            var body = await context.Body();
            var username = body.Str("username");
            var password = body.Str("password");

            if (string.IsNullOrWhiteSpace(username))
                return Results.BadRequest(new { error = "Nama pengguna wajib diisi." });

            using var connection = Db.Open();

            var user = Sql.Row(connection,
                @"SELECT u.id, u.username, u.password_hash, u.display_name, u.role, u.is_active,
                         t.id AS tenant_id, t.name AS tenant_name
                    FROM users u
                    JOIN tenants t ON t.id = u.tenant_id
                   WHERE u.username = @p0",
                username);

            // Pesan yang sama untuk pengguna tak dikenal dan kata sandi salah.
            // Pesan yang berbeda memberi tahu penebak nama mana yang benar-benar ada.
            var invalid = Results.Json(
                new { error = "Nama pengguna atau kata sandi salah." },
                statusCode: StatusCodes.Status401Unauthorized);

            if (user == null) return invalid;
            if (Convert.ToInt64(user["isActive"]) == 0) return invalid;
            if (!Passwords.Verify(password, (string)user["passwordHash"])) return invalid;

            var principal = new Principal
            {
                UserId = (string)user["id"],
                Username = (string)user["username"],
                TenantId = (string)user["tenantId"],
                TenantName = (string)user["tenantName"],
                Role = (string)user["role"],
                DisplayName = (string)user["displayName"],
            };

            Sql.Exec(connection, "UPDATE users SET last_login_at = @p0 WHERE id = @p1",
                Sql.Now(), principal.UserId);

            return Results.Ok(new
            {
                token = Tokens.Issue(principal),
                expiresInSeconds = (int)Tokens.Lifetime.TotalSeconds,
                user = new
                {
                    principal.Username,
                    principal.DisplayName,
                    principal.Role,
                    tenant = principal.TenantName,
                },
            });
        });

        // --------------------------------------------------------------
        app.MapGet("/api/auth/me", (HttpContext context) =>
        {
            var principal = context.User();

            return Results.Ok(new
            {
                principal.Username,
                principal.DisplayName,
                principal.Role,
                tenant = principal.TenantName,
            });
        });

        // --------------------------------------------------------------
        app.MapPost("/api/auth/password", async (HttpContext context) =>
        {
            var body = await context.Body();
            var current = body.Str("currentPassword");
            var next = body.Str("newPassword");

            if (string.IsNullOrEmpty(next) || next.Length < 6)
                return Results.BadRequest(new { error = "Kata sandi baru minimal 6 karakter." });

            using var connection = Db.Open();

            var stored = (string)Sql.Scalar(connection,
                "SELECT password_hash FROM users WHERE id = @p0", context.User().UserId);

            if (!Passwords.Verify(current, stored))
                return Results.BadRequest(new { error = "Kata sandi sekarang salah." });

            Sql.Exec(connection, "UPDATE users SET password_hash = @p0 WHERE id = @p1",
                Passwords.Hash(next), context.User().UserId);

            return Results.Ok(new { ok = true });
        });
    }
}

File: ForgeHub/backend/src/main/java/id/jakforge/forgehub/api/AuthApi.java
package id.jakforge.forgehub.api;

import id.jakforge.forgehub.data.Db;
import id.jakforge.forgehub.security.CurrentUser;
import id.jakforge.forgehub.security.ForgeHubPrincipal;
import id.jakforge.forgehub.security.JwtService;
import id.jakforge.forgehub.security.Passwords;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Masuk, siapa saya, dan ganti kata sandi.
 *
 * <p>Bentuk jawabannya mengikuti ForgeHub .NET persis. Yang benar-benar dibaca
 * klien cuma satu medan — {@code token} — tapi sisanya dipakai dasbor, dan
 * medan yang hilang di sana muncul sebagai "undefined" di layar, bukan sebagai
 * kesalahan yang kelihatan.
 */
@RestController
@RequestMapping("/api")
public class AuthApi {

    private final Db db;
    private final JwtService jwt;

    public AuthApi(Db db, JwtService jwt) {
        this.db = db;
        this.jwt = jwt;
    }

    @GetMapping("/health")
    public Map<String, Object> health() {
        Map<String, Object> hasil = new LinkedHashMap<>();
        hasil.put("product", "ForgeHub");
        hasil.put("status", "OK");
        hasil.put("time", Db.nowText());
        return hasil;
    }

    @PostMapping("/auth/login")
    public ResponseEntity<?> login(@RequestBody(required = false) Map<String, Object> body) {
        String username = Badan.teks(body, "username");
        String password = Badan.teks(body, "password");

        if (username == null || username.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("error", "Nama pengguna wajib diisi."));
        }

        Map<String, Object> user = db.row("""
                SELECT u.id, u.username, u.password_hash, u.display_name, u.role, u.is_active,
                       t.id AS tenant_id, t.name AS tenant_name
                  FROM users u
                  JOIN tenants t ON t.id = u.tenant_id
                 WHERE u.username = ?
                """, username);

        // Jawaban yang SAMA untuk pengguna tak dikenal dan kata sandi salah.
        // Jawaban yang berbeda memberi tahu penebak nama mana yang benar-benar
        // ada, dan itu memisahkan satu tebakan menjadi dua yang jauh lebih murah.
        ResponseEntity<?> ditolak = ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(Map.of("error", "Nama pengguna atau kata sandi salah."));

        if (user == null) return ditolak;
        if (!Boolean.TRUE.equals(user.get("isActive"))) return ditolak;
        if (!Passwords.verify(password, (String) user.get("passwordHash"))) return ditolak;

        UUID userId = Db.uuid((String) user.get("id"));
        UUID tenantId = Db.uuid((String) user.get("tenantId"));
        String role = (String) user.get("role");

        db.exec("UPDATE users SET last_login_at = now() WHERE id = ?", userId);

        // Kata sandi yang tersimpan dengan putaran lebih sedikit di-hash ulang
        // SEKARANG, saat kata sandi polosnya ada di tangan. Ini satu-satunya
        // saat itu mungkin; sesudah ini yang tersimpan hanya hash-nya.
        String tersimpan = (String) user.get("passwordHash");
        if (Passwords.needsRehash(tersimpan)) {
            db.exec("UPDATE users SET password_hash = ? WHERE id = ?", Passwords.hash(password), userId);
        }

        Map<String, Object> hasil = new LinkedHashMap<>();
        hasil.put("token", jwt.issue(userId, tenantId, username, role));
        hasil.put("expiresInMinutes", jwt.getExpirationMinutes());
        hasil.put("userId", userId.toString());
        hasil.put("username", user.get("username"));
        hasil.put("displayName", user.get("displayName"));
        hasil.put("role", role);
        hasil.put("tenantId", tenantId.toString());
        hasil.put("tenantName", user.get("tenantName"));

        return ResponseEntity.ok(hasil);
    }

    @GetMapping("/auth/me")
    public ResponseEntity<?> me() {
        ForgeHubPrincipal p = CurrentUser.get();

        Map<String, Object> user = db.row("""
                SELECT u.id, u.username, u.display_name, u.email, u.role, u.is_active,
                       u.created_at, u.last_login_at, t.name AS tenant_name
                  FROM users u
                  JOIN tenants t ON t.id = u.tenant_id
                 WHERE u.id = ? AND u.tenant_id = ?
                """, p.userId(), p.tenantId());

        if (user == null) {
            // Token sah untuk pengguna yang sudah tidak ada. Itu bukan 500:
            // yang salah adalah tokennya, bukan servernya.
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("error", "Pengguna sudah tidak ada."));
        }

        return ResponseEntity.ok(user);
    }

    @PostMapping("/auth/password")
    public ResponseEntity<?> gantiSandi(@RequestBody(required = false) Map<String, Object> body) {
        ForgeHubPrincipal p = CurrentUser.get();

        String lama = Badan.teks(body, "currentPassword");
        String baru = Badan.teks(body, "newPassword");

        if (baru == null || baru.length() < 8) {
            return ResponseEntity.badRequest()
                    .body(Map.of("error", "Kata sandi baru minimal 8 karakter."));
        }

        Object tersimpan = db.scalar("SELECT password_hash FROM users WHERE id = ?", p.userId());

        // Kata sandi lama tetap diminta walau penggunanya sudah membawa token
        // yang sah. Token bisa berasal dari layar yang ditinggal terbuka;
        // kata sandi lama hanya diketahui pemiliknya.
        if (!Passwords.verify(lama, (String) tersimpan)) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("error", "Kata sandi saat ini salah."));
        }

        db.exec("UPDATE users SET password_hash = ? WHERE id = ?", Passwords.hash(baru), p.userId());

        return ResponseEntity.ok(Map.of("status", "OK"));
    }
}
